### PR TITLE
REST > Get Transactions 

### PR DESCRIPTION
**Pull request by Ninjeneer**
In OAS doc, the `ChargeBoxIDs` parameters is only a temporary fix to avoid changing or breaking front-end/mobile, as on server-side we use sometimes "ChargingStationID" or "ChargeBoxID". 
I will standardize those parameters names when deploying API changes to front-end and mobile, and remove the `ChargeBoxIDs` in OAS file.

### CODE REVIEW

File: src/server/rest/v1/service/TransactionService.ts [hunk 1]
```diff
@@ -40,6 +40,28 @@ import moment from 'moment-timezone';
...
+    const filteredRequest = TransactionSecurity.filterGetTransactions(req.query);
+    let transactions: DataResult<Transaction>;
+    if (filteredRequest.Status) {
+      if (filteredRequest.Status === 'completed') {
+        transactions = await TransactionService.getTransactionsCompleted(ServerAction.TRANSACTIONS_COMPLETED, req);
+      } else if (filteredRequest.Status === 'active') {
+        transactions = await TransactionService.getTransactionsActive(ServerAction.TRANSACTIONS_ACTIVE, req);
+      }
+    } else {
+      const completed = await TransactionService.getTransactionsCompleted(ServerAction.TRANSACTIONS_COMPLETED, req);
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TransactionService.ts:54]: This does not work.
You break the pagination.

File: src/server/rest/v1/service/security/TransactionSecurity.ts [hunk 2]
```diff
@@ -110,4 +110,10 @@ export default class TransactionSecurity {
     }
     return filteredRequest;
   }
+
+  public static filterGetTransactions(request: any): HttpGetTransactionsRequest {
+    return {
+      Status: sanitize(request.Status),
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/security/TransactionSecurity.ts:116]: Should be part of the big http transaction type

File: src/server/rest/v1/service/TransactionService.ts [hunk 3]
```diff
@@ -1071,4 +1085,20 @@ export default class TransactionService {
     );
     return transactions;
   }
+
+  private static async getTransactionsCompleted(action: ServerAction, req: Request): Promise<DataResult<Transaction>> {
+    return await TransactionService.getTransactions(req, action, { completedTransactions: true }, [
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TransactionService.ts:1090]: Set the query string Status param to 'completed' and call the getTransactions() in which you can provide a IF to add the proper filter (if Status is not provided return all transactions)

File: src/server/rest/v1/service/TransactionService.ts [hunk 4]
```diff
@@ -1071,4 +1085,20 @@ export default class TransactionService {
...
+      'id', 'chargeBoxID', 'timestamp', 'issuer', 'stateOfCharge', 'timezone', 'connectorId', 'meterStart', 'siteAreaID', 'siteID',
+      'stop.roundedPrice', 'stop.price', 'stop.priceUnit', 'stop.inactivityStatus', 'stop.stateOfCharge', 'stop.timestamp', 'stop.totalConsumptionWh',
+      'stop.totalDurationSecs', 'stop.totalInactivitySecs', 'stop.meterStop', 'billingData.invoiceID', 'ocpi', 'ocpiWithCdr', 'tagID', 'stop.tagID',
+    ]);
+  }
+
+  private static async getTransactionsActive(action: ServerAction, req: Request): Promise<DataResult<Transaction>> {
+    return await TransactionService.getTransactions(req, action, { completedTransactions: false }, [
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TransactionService.ts:1098]: Set the query string Status param to 'active' and call the getTransactions() in which you can provide a IF to add the proper filter (if Status is not provided return all transactions)

File: src/server/rest/v1/service/TransactionService.ts [hunk 5]
```diff
@@ -685,22 +707,14 @@ export default class TransactionService {
   }
 
   public static async handleGetTransactionsActive(action: ServerAction, req: Request, res: Response, next: NextFunction): Promise<void> {
-    const transactions = await TransactionService.getTransactions(req, action, { completedTransactions: false }, [
-      'id', 'chargeBoxID', 'timestamp', 'issuer', 'stateOfCharge', 'timezone', 'connectorId', 'status', 'meterStart', 'siteAreaID', 'siteID',
-      'currentTotalDurationSecs', 'currentTotalInactivitySecs', 'currentInstantWatts', 'currentTotalConsumptionWh', 'currentStateOfCharge',
-      'currentCumulatedPrice', 'currentInactivityStatus', 'roundedPrice', 'price', 'priceUnit', 'tagID',
-    ]);
+    const transactions = await TransactionService.getTransactionsActive(action, req);
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TransactionService.ts:710]: No need to call a private method here if the change is done above.

File: src/server/rest/v1/service/TransactionService.ts [hunk 6]
```diff
@@ -685,22 +707,14 @@ export default class TransactionService {
...
   }
 
   public static async handleGetTransactionsCompleted(action: ServerAction, req: Request, res: Response, next: NextFunction): Promise<void> {
     // Get transaction
-    const transactions = await TransactionService.getTransactions(req, action, { completedTransactions: true }, [
-      'id', 'chargeBoxID', 'timestamp', 'issuer', 'stateOfCharge', 'timezone', 'connectorId', 'meterStart', 'siteAreaID', 'siteID',
-      'stop.roundedPrice', 'stop.price', 'stop.priceUnit', 'stop.inactivityStatus', 'stop.stateOfCharge', 'stop.timestamp', 'stop.totalConsumptionWh',
-      'stop.totalDurationSecs', 'stop.totalInactivitySecs', 'stop.meterStop', 'billingData.invoiceID', 'ocpi', 'ocpiWithCdr', 'tagID', 'stop.tagID',
-    ]);
+    const transactions = await TransactionService.getTransactionsCompleted(action, req);
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TransactionService.ts:717]: No need to call a private method here if the change is done above.

File: src/server/rest/v1/service/TransactionService.ts [hunk 7]
```diff
@@ -40,6 +40,28 @@ import moment from 'moment-timezone';
 const MODULE_NAME = 'TransactionService';
 
 export default class TransactionService {
+
+  public static async handleGetTransactions(action: ServerAction, req: Request, res: Response, next: NextFunction): Promise<void> {
+    const filteredRequest = TransactionSecurity.filterGetTransactions(req.query);
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/TransactionService.ts:45]: Call right await the getTransactions() with the proper fields projection (related to my other comments below)

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]

File: src/server/rest/v1/service/TransactionService.ts (hunk 4 above)
Comment (Ninjeneer) [reply src/server/rest/v1/service/TransactionService.ts:1098]: I understand how to retrieve completed or active transactions from the same getTransactions() method. However I can't figure out how to return both from the same call, as they need different projections, without breaking pagination.

File: src/server/rest/v1/service/security/TransactionSecurity.ts (hunk 2 above)
Comment (Ninjeneer) [reply src/server/rest/v1/service/security/TransactionSecurity.ts]: Done

File: src/server/rest/v1/service/TransactionService.ts (hunk 3 above)
Comment (Ninjeneer) [reply src/server/rest/v1/service/TransactionService.ts]: Done

File: src/server/rest/v1/service/TransactionService.ts (hunk 4 above)
Comment (Ninjeneer) [reply src/server/rest/v1/service/TransactionService.ts]: Done

File: src/server/rest/v1/service/TransactionService.ts (hunk 5 above)
Comment (Ninjeneer) [reply src/server/rest/v1/service/TransactionService.ts]: Removed private method

File: src/server/rest/v1/service/TransactionService.ts (hunk 6 above)
Comment (Ninjeneer) [reply src/server/rest/v1/service/TransactionService.ts]: Removed private method

File: src/server/rest/v1/service/TransactionService.ts (hunk 7 above)
Comment (Ninjeneer) [reply src/server/rest/v1/service/TransactionService.ts]: Done

File: src/server/rest/v1/service/TransactionService.ts (hunk 1 above)
Comment (Ninjeneer) [reply src/server/rest/v1/service/TransactionService.ts]: Removed

**Review by LucasBrazi06 — APPROVED** [review]
GUT !!!

**Pull request merged**